### PR TITLE
Cache indexer service account credentials (#2056)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
> This looks good to me, but consider moving the funciton to `azul.deployment`. If feels like it belongs as a @cached_property of a GCP equivalent of the AWS class there, but that may be beyond the scope of your PR.
> 
> Maybe ask @hannes-ucsc and see what he prefers.

@hannes-ucsc Thoughts on this? We didn't decide to move the function when we began it using for TDR, so I'm not sure.

File: src/azul/dss.py [hunk 1]
```diff
@@ -220,6 +223,14 @@ def __init__(self, *args, num_workers: int = None, **kwargs):
                          **kwargs)
 
 
+# Cached to prevent exceeding requests limits during TDR reindex
+@lru_cache
+def _google_service_account_key() -> str:
```
Comment (hannes-ucsc) [inline src/azul/dss.py:228]:
```suggestion
def _service_account_creds() -> str:
```

File: src/azul/dss.py [hunk 2]
```diff
@@ -220,6 +223,14 @@ def __init__(self, *args, num_workers: int = None, **kwargs):
...
+    sm = boto3.client('secretsmanager')
+    creds = sm.get_secret_value(SecretId=config.secrets_manager_secret_name('indexer', 'google_service_account'))
+    return creds['SecretString']
```
Comment (hannes-ucsc) [inline src/azul/dss.py:231]:
```suggestion
    return creds
```

File: src/azul/dss.py [hunk 3]
```diff
@@ -220,6 +223,14 @@ def __init__(self, *args, num_workers: int = None, **kwargs):
                          **kwargs)
 
 
+# Cached to prevent exceeding requests limits during TDR reindex
+@lru_cache
+def _google_service_account_key() -> str:
+    sm = boto3.client('secretsmanager')
+    creds = sm.get_secret_value(SecretId=config.secrets_manager_secret_name('indexer', 'google_service_account'))
```
Comment (hannes-ucsc) [inline src/azul/dss.py:230]: wrap this line a bit.

File: src/azul/dss.py [hunk 4]
```diff
@@ -228,10 +241,9 @@ def shared_credentials():
     credentials with write access to DSS aren't available or b) you want to act on behalf of the Azul indexer,
     or rather *as* the indexer.
     """
-    sm = boto3.client('secretsmanager')
-    creds = sm.get_secret_value(SecretId=config.secrets_manager_secret_name('indexer', 'google_service_account'))
+    google_key = _service_account_creds()['SecretString']
```
Comment (hannes-ucsc) [inline src/azul/dss.py:244]:
```suggestion
    secret = _service_account_creds()['SecretString']
```

**Comment by hannes-ucsc** [comment]
We'll move this module later. You made the right choice of not moving it in this PR.

File: src/azul/dss.py [hunk 5]
```diff
@@ -220,6 +223,16 @@ def __init__(self, *args, num_workers: int = None, **kwargs):
                          **kwargs)
 
 
+# Cached to prevent exceeding requests limits during TDR reindex
```
Comment (hannes-ucsc) [inline src/azul/dss.py:226]: That's a bit of in implementation detail of the call site. The reason for caching is universal: a resources is expensive to create and frequently used.

File: src/azul/dss.py [hunk 6]
```diff
@@ -220,6 +223,16 @@ def __init__(self, *args, num_workers: int = None, **kwargs):
...
+@lru_cache
+def _service_account_creds() -> JSON:
+    sm = boto3.client('secretsmanager')
+    creds = sm.get_secret_value(
+        SecretId=config.secrets_manager_secret_name('indexer', 'google_service_account')
```
Comment (hannes-ucsc) [inline src/azul/dss.py:231]: The secret name should have been a parameter.

**Comment by hannes-ucsc** [comment]
I'm just gonna make these changes.

**Pull request merged**